### PR TITLE
949 - Included the cron to parse the data endpoints and to push it to solr.

### PR DESCRIPTION
**Pull request by riginoommen**
# Resolves 949

### 

# Explain the feature
Support to read the config on mapping fields from search response and push to solr. 

- APIs to manage mapping configuration
- Cron to pull data and manipulate it as per the preferences
- Push the data to the solr.

## Does this PR introduce a breaking change
Yes

## Screenshots
N/A

### Ready-for-merge Checklist

- [x] Expected files: all files in this pull request are related to one feature request or issue (no stragglers)?
- [x] Does the change have appropriate unit tests?
- [x] Did tests pass?
- [x] Did you update or add any necessary documentation (README.md, WHY.md, etc.)?
- [x] Was this feature demo'd and the design review approved?

### CODE REVIEW

**Comment by riginoommen** [comment]
You can still do a code-review for this PR. Because i made this as a draft PR because needs some more testing

File: packages/search-service/service.ts [hunk 1]
```diff
@@ -54,3 +86,10 @@ apollo.installSubscriptionHandlers(server);
 export default server.listen({ port: port }, () => {
   console.log(`Microservice running on ${process.env.NODE_ENV} at ${port}${apollo.graphqlPath}`);
 });
+
+// Cron to sync the search data.
+schedule.scheduleJob( '0 0 * * *', () => {
+  console.log( `Running Search sync cron on ${ new Date().toISOString() }` );
+  const searchMapCron = new SearchMapCron();
+  searchMapCron.searchMapTrigger();
+});
```
Comment (deshmukhmayur) [inline packages/search-service/service.ts:95]:
```suggestion
});

```

File: packages/search-service/src/search-mapping/cron.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                documents: _.uniqBy(mappedList, 'id')
+            };
+            const response = await SearchIndexHelper.index(searchInput);
+            statusList.push({ status: response.status });
+            if (searchConfigList.length === configIndex + 1) {
+                console.info('Index Status -', statusList);
+            }
+        });
+    }
+}
```
Comment (deshmukhmayur) [inline packages/search-service/src/search-mapping/cron.ts:40]:
```suggestion
}

```

File: packages/search-service/src/types.d.ts [hunk 3]
```diff
@@ -38,4 +38,37 @@ type SearchResponseType = {
...
+    preferences: {
+        iconUrl: string,
+        searchUrlTemplate: string
+        searchUrlParams: string[]
+    },
+    createdBy: string;
+    createdOn: Date;
+    updatedBy: string;
+    updatedOn: Date;
+}
```
Comment (deshmukhmayur) [inline packages/search-service/src/types.d.ts:74]:
```suggestion
}

```

File: packages/search-service/src/search-mapping/typedef.graphql [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    createSearchMap(input: SearchMapInput): SearchMapType
+    """
+    update the Search Map configuration in database
+    """
+    updateSearchMap(input: SearchMapInput): SearchMapType
+    """
+    Delete Search Map configuration from database
+    """
+    deleteSearchMap(_id: String!): SearchMapType
+}
```
Comment (deshmukhmayur) [inline packages/search-service/src/search-mapping/typedef.graphql:99]:
```suggestion
}

```

File: packages/search-service/src/search-mapping/helpers.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        return object;
+    }
+
+    public fetchApi(apiUrl: any, query: any, variables: any, mode: any) {
+        let headers = new Headers();
+        let body = JSON.stringify({
+            query,
+            variables
+        });
+        headers.append(`Authorization`, `${process.env.GATEWAY_AUTH_TOKEN}`);
```
Comment (deshmukhmayur) [inline packages/search-service/src/search-mapping/helpers.ts:34]: You cannot use the Gateway Auth token here, the Gateway Auth toke only works for One Platform API Gateway. the user can have their APIs hosted outside the API Gateway, so we need to get the Authorization Header value from the user and store it in the config.

File: packages/search-service/src/search-mapping/schema.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { Document, Model, model, Schema } from 'mongoose';
+
+export const SearchMapSchema: Schema = new Schema({
+  name: String,
```
Comment (deshmukhmayur) [inline packages/search-service/src/search-mapping/schema.ts:4]: Add another field for SPA. That should store the reference to the SPA id from the Apps/Home Service.
```suggestion
  name: String,
  spa: { type: String, },
```

File: packages/search-service/src/search-mapping/schema.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  fields: [{
+    from: String,
+    to: String
+  }],
+  apiConfig: {
+    model: { type: String, enum: ['GRAPHQL', 'REST'] },
+    mode: String,
+    apiUrl: String,
+    param: String,
+    gqlQuery: String
```
Comment (deshmukhmayur) [inline packages/search-service/src/search-mapping/schema.ts:14]: Add support for user defined Authorization header, and other custom headers if their API has some special header requirements to function properly.

Also make the necessary changes in the helper functions.
```suggestion
    gqlQuery: String,
    authorizationHeader: { type: String, },
    customHeaders: [{ key: String, value: String, }],
```

File: packages/search-service/src/search-mapping/schema.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import { Document, Model, model, Schema } from 'mongoose';
+
+export const SearchMapSchema: Schema = new Schema({
+  name: String,
+  fields: [{
+    from: String,
+    to: String
+  }],
+  apiConfig: {
+    model: { type: String, enum: ['GRAPHQL', 'REST'] },
```
Comment (deshmukhmayur) [inline packages/search-service/src/search-mapping/schema.ts:10]: Can you make this more generic?

i.e. No need to ask for GraphQL / ReST. Just ask for request method (GET / POST), and the request body if POST is selected. A GraphQL request is essentially a POST request with a JSON body, so the graphql query can also be provided as `body`.

Comment (deshmukhmayur) [reply packages/search-service/src/search-mapping/schema.ts:10]: We can simplify it further on the UI using a Postman-like UI for GraphQL requests, while storing it as a normal POST request in the backend.

**Review by deshmukhmayur — CHANGES_REQUESTED** [review]

**Review by deshmukhmayur — APPROVED** [review]

File: packages/search-service/src/search-mapping/helpers.ts [hunk 9]
```diff
@@ -38,7 +38,7 @@ class APITemplateHelper {
         } );
         headers.append( `Authorization`, `${ process.env.GATEWAY_AUTH_TOKEN }` );
         headers.append( `Content-Type`, `application/json` );
-        return fetch(`${process.env.API_GATEWAY}`, {
+        return fetch(`http://api-gateway:8080/graphql`, {
```
Comment (deshmukhmayur) [inline packages/search-service/src/search-mapping/helpers.ts:41]: This should use the environment variable. Don't hard code the urls, even for local testing.
```suggestion
        return fetch(process.env.API_GATEWAY, {
```

**Review by deshmukhmayur — CHANGES_REQUESTED** [review]

File: packages/search-service/src/search-mapping/helpers.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+                        apps {
+                            id
+                            name
+                        }
+                    }`,
+            variables: {}
+        } );
+        headers.append( `Authorization`, `${ process.env.API_GATEWAY }` );
+        headers.append( `Content-Type`, `application/json` );
+        return fetch( `${ process.env.GATEWAY_AUTH_TOKEN }`, {
```
Comment (deshmukhmayur) [inline packages/search-service/src/search-mapping/helpers.ts:41]:
```suggestion
        headers.append( `Authorization`, `${ process.env.GATEWAY_AUTH_TOKEN }` );
        headers.append( `Content-Type`, `application/json` );
        return fetch( `${ process.env.API_GATEWAY }`, {
```

**Review by deshmukhmayur — CHANGES_REQUESTED** [review]

File: packages/search-service/src/search-mapping/helpers.ts (hunk 10 above)
Comment (riginoommen) [reply packages/search-service/src/search-mapping/helpers.ts]: Updated

**Review by deshmukhmayur — APPROVED** [review]

**Review by ghanlohar — APPROVED** [review]
Looks good to me.

**Pull request merged**